File: python/count_of_smaller_numbers_after_self.py
class FenwickTree(object):
    def __init__(self, n):
        self.sum_array = [0] * (n + 1)
        self.n = n

    def lowbit(self, x):
        return x & (-x)

    def add(self, i, delta):
        while i < self.n:
            self.sum_array[i] += delta
            i += self.lowbit(i)

    def sum(self, i):
        res = 0
        while i > 0:
            res += self.sum_array[i]
            i -= self.lowbit(i)
        return res


class Node(object):
    def __init__(self, n):
        self.num = n
        self.cnt = 0


class Solution(object):
    def __megre(self, left, right):
        res = []
        lp = 0
        rp = 0
        while lp < len(left) and rp < len(right):
            if left[lp].num <= right[rp].num:
                left[lp].cnt += rp
                res.append(left[lp])
                lp += 1
            else:
                res.append(right[rp])
                rp += 1

        while lp < len(left):
            left[lp].cnt += rp
            res.append(left[lp])
            lp += 1

        while rp < len(right):
            res.append(right[rp])
            rp += 1

        return res

    def __sort(self, nlist):
        if len(nlist) <= 1:
            return nlist
        else:
            mid = len(nlist) / 2
            left = self.__sort(nlist[0:mid])
            right = self.__sort(nlist[mid:])
            return self.__megre(left, right)

    def countSmaller0(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        nlist = []
        for i in nums:
            nlist.append(Node(i))
        self.__sort(nlist)

        return [x.cnt for x in nlist]

    def countSmaller(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        d = {}
        for i, num in enumerate(sorted(set(nums))):
            d[num] = i + 1

        tree = FenwickTree(len(nums))
        res = [0] * len(nums)
        for i in xrange(len(nums)-1, -1, -1):
            res[i] = tree.sum(d[nums[i]] - 1)
            tree.add(d[nums[i]], 1)

        return res


if __name__ == "__main__":
    s = Solution()
    print s.countSmaller([5, 2, 6, 1])

File: cpp/src/count_of_smaller_numbers_after_self.cpp
#include "leetcode.h"
#include <vector>
#include <map>
#include <algorithm>

using namespace std;

class FenwickTree {
 private:
  int n;
  vector<int> sumArray;

  inline int lowbit(int x) {
    return x & -x;
  }

 public:
  FenwickTree(int n): n(n), sumArray(n + 1, 0) {}

  void add(int x, int val) {
    while (x <= this->n) {
      this->sumArray[x] += val;
      x += this->lowbit(x);
    }
  }

  int sum(int x) {
    int res = 0;
    while (x > 0) {
      res += sumArray[x];
      x -= this->lowbit(x);
    }
    return res;
  }
};

class Solution {
 private:
  map<int, int> orderDict(vector<int> & nums) {
    map<int, int> dict;
    vector<int> sorted(nums);
    sort(sorted.begin(), sorted.end());
    int order = 1;
    int prev = sorted[0];
    dict.insert(pair<int, int>(prev, order));
    for (size_t i = 1; i < sorted.size(); i++) {
      if (sorted[i] != prev) {
        order += 1;
        prev = sorted[i];
        dict[prev] = order;
      }
    }
    return dict;
  }

 public:
  vector<int> countSmaller(vector<int>& nums) {
    vector<int> res(nums.size(), 0);
    if (nums.size() == 0) {
      return res;
    }

    map<int, int> dict = this->orderDict(nums);
    FenwickTree tree(nums.size());
    for (int i = nums.size() - 1; i >= 0; i--) {
      res[i] = tree.sum(dict[nums[i]] - 1);
      tree.add(dict[nums[i]], 1);
    }
    return res;
  }
};

int main(int argc, char **argv) {
  vector<int> nums = {5, 2, 1, 2, 6, 1};
  Solution s;
  vector<int> res = s.countSmaller(nums);
  for (vector<int>::iterator it = res.begin(); it != res.end(); it++) {
    debug_log("%d\n", *it);
  }
  return 0;
}
